src/expand.rs

```diff
@@ -0,0 +1,175 @@
+    ///
+    /// Expands all instances of `\num` or `\g<name>` in `replacement`
+    /// to the corresponding capture group `num` or `name`, and writes
+    /// them to the `dst` buffer given.
+    ///
+    /// `name` may be an integer corresponding to the index of the
+    /// capture group (counted by order of opening parenthesis where `\0` is the
+    /// entire match) or it can be a name (consisting of letters, digits or
+    /// underscores) corresponding to a named capture group.
+    ///
+    /// `num` must be an integer corresponding to the index of the
+    /// capture group.
+    ///
+    /// If `num` or `name` isn't a valid capture group (whether the name doesn't exist
+    /// or isn't a valid index), then it is replaced with the empty string.
+    ///
+    /// The longest possible number is used. e.g., `\10` looks up capture
+    /// group 10 and not capture group 1 followed by a literal 0.
+    ///
+    /// To write a literal `\`, use `\\`.
+    pub fn python() -> Self {
+        Expander {
+            sub_char: '\\',
+            open: "g<",
+            close: ">",
+            allow_undelimited_name: false,
+            strict: false,
+        }
+    }
+
+    /// Sets whether this expander will report errors caused by unknown
+    /// group names and unclosed substitution expressions.
+    ///
+    /// Expanders are non-strict by default.
+    pub fn set_strict(&mut self, value: bool) {
+        self.strict = value;
+    }
+
+    /// Quotes the substitution character in `text` so it appears literally
+    /// in the output of `expansion`.
+    ///
+    /// ```
+    /// assert_eq!(
+    ///     fancy_regex::Expander::default().quote("Has a literal $ sign."),
+    ///     "Has a literal $$ sign.",
+    /// );
+    /// ```
+    pub fn quote<'a>(&self, text: &'a str) -> Cow<'a, str> {
+        if text.contains(self.sub_char) {
+            let mut quoted = String::with_capacity(self.sub_char.len_utf8() * 2);
+            quoted.push(self.sub_char);
+            quoted.push(self.sub_char);
+            Cow::Owned(text.replace(self.sub_char, &quoted))
+        } else {
+            Cow::Borrowed(text)
+        }
+    }
+
+    /// Expands the template string `template` using the syntax defined
+    /// by this expander and the values of capture groups from `captures`.
+    ///
+    /// Always succeeds when this expander is not strict.
+    pub fn expansion<'t>(&self, captures: &Captures<'t>, template: &str) -> io::Result<String> {
+        let mut cursor = io::Cursor::new(Vec::new());
+        self.write_expansion(&mut cursor, captures, template)?;
+        Ok(String::from_utf8(cursor.into_inner()).expect("expansion is UTF-8"))
+    }
+
+    /// Appends the expansion produced by `expansion` to `dst`.  Possibly more efficient
+    /// than calling `expansion` directly.
+    pub fn append_expansion<'t>(
+        &self,
+        dst: &mut String,
+        captures: &Captures<'t>,
+        template: &str,
```

Comment:
I like the new names :)!

One more comment on the API: Should we put the `template` arguments before `captures`?

Why? Because that would be similar to how `write!` is used, e.g.: `write!(f, "Regex error: {}", e)`. So, destination first, then template string, then arguments (which in our case come from the captures). What do you think?

Reply:
Seems reasonable to me.